Variable number of G29 grid points (UBL, Bilinear)

lukasradek opened the pull request:
### Description
Allows for variable number of grid probing points up to the number (per axis) specified in config file.

The logic to determine the number of grid points can be based on many things (menu config, GCODE,...), but for now the aim is to automatically decrease the number of points when probing small area and point spacing is smaller than specified in `Configuration.h` by `GRID_MIN_SPACING`, which seems to be the issue that most people have with number of probing points.

### Requirements
Probe, G29 Bilinear Leveling

### Benefits
It addresses a lot of user requests and also decreases the  probing time on larger machines when printing smaller things.
It goes together well with G29 LRFB syntax.

### Configurations
The proposed changes in `Configuration.h` are submited in the PR.

### Related Issues
- https://github.com/MarlinFirmware/Marlin/issues/10547
- https://github.com/MarlinFirmware/Marlin/issues/25660
- https://github.com/MarlinFirmware/Marlin/issues/25689
- https://github.com/MarlinFirmware/Marlin/issues/22861
- https://github.com/MarlinFirmware/Marlin/issues/19209
- https://github.com/MarlinFirmware/Marlin/issues/20974

### Work In Progress

**Todo Minimums:**

- [x] Implement for G29 H syntax
- [x] Extend EEPROM storage logic
- [x] Modify Bilinear Subdivision - `subdivide_mesh`
- [x] Modify Extrapolation for non-rectangular beds - `extrapolate_unprobed_bed_level`

**ToDo Possibilities:**

- [x] Accept arguments from G29 to set number of points directly
- [x] Let user specify spacing (instead of grid points) in config and let the firmware determine number of points from probing area size. 
- [ ] Allow 2x2 grids in `Configuration.h` (is there any reason why there is a sanity check that requires min 3x3?
- [x] VARIABLE_GRID_POINTS flag should be automatically enabled when AUTO_BED_LEVELING_LINEAR is enabled. At this point the code always checks for #if ANY(AUTO_BED_LEVELING_LINEAR , VARIABLE_GRID_POINTS), but linear leveling has variable grid points actually implemented by default. Is there anything that such change would break?

**In its current implementation Variable Grid Points work (math algos need to be checked) unless you want to use features from unchecked files mentioned below.**

### Extending Support and ToDo Files 

Files that needs to be adjusted in order to fully support Variable Grid Points.
Without those adjustments, they should just fail gracefully and display NaN when showing value of grid point beyond the variable grid point limit (most of those changes concern UI printing leveling grid).

**MESH_EDIT_MENU Support**
- [x] lcd\menu\menu_bed_leveling.cpp

**G-code support**
- [x] gcode\bedlevel\G26.cpp
- [x] gcode\bedlevel\G42.cpp
- [x] gcode\bedlevel\M420.cpp
- [x] gcode\bedlevel\abl\M421.cpp

**Utils**
- [x] inc\Conditionals_LCD.h

**e3v2 Support**
- [ ] lcd\e3v2\jyersui\dwin.cpp
- [ ] lcd\e3v2\proui\bedlevel_tools.cpp
- [ ] lcd\e3v2\proui\dwin.cpp
- [ ] lcd\e3v2\proui\meshviewer.cpp

**ExtUI Support**
- [ ] lcd\extui\ui_api.cpp
- [ ] lcd\extui\anycubic_chiron\chiron_tft.cpp
- [ ] lcd\extui\anycubic_vyper\dgus_tft.cpp
- [ ] lcd\extui\dgus\mks\DGUSScreenHandler.cpp
- [ ] lcd\extui\dgus_e3s1pro\DGUSTxHandler.cpp
- [ ] lcd\extui\dgus_reloaded\DGUSScreenHandler.cpp
- [ ] lcd\extui\dgus_reloaded\DGUSTxHandler.cpp
- [ ] lcd\extui\ftdi_eve_touch_ui\generic\bed_mesh_base.cpp
- [ ] lcd\extui\ftdi_eve_touch_ui\generic\bed_mesh_edit_screen.cpp
- [ ] lcd\extui\ftdi_eve_touch_ui\generic\bed_mesh_view_screen.cpp

It looks like IA_CREALITY implements its own leveling and thus doesn't bother with `VARIBALE_GRID_POINTS`, but just to be sure...
- [ ] lcd\extui\ia_creality\ia_creality_extui.cpp
- [ ] lcd\extui\ia_creality\ia_creality_rts.cpp

### Asking for HELP

Verify that mathematical algorithms in `feature/bedlevel/alb/bbl.cpp` class `LevelingBilinear` are correct.
- [ ] Extrapolation - `extrapolate_unprobed_bed_level` and `extrapolate_one_point`
- [ ] Cartesian moves - `line_to_destination.`
- [ ] Bilinear Subdivision - `virt_2cmr`, `virt_cmr`, `virt_coord` 

Verify that EEPROM data storing and retrieval is implemented correctly
- [ ] EEPROM read a write of `bedlevel.nr_grid_points` is correct

Implement **e3v2** and **ExtUI** support since I don't have the hardware to test it with.
- [ ] e3v2
- [ ] ExtUI

lukasradek committed:
Basic implementation

Works for Bilinear G29 LRFB syntax with square beds and without BILINEAR_SUBDIVISION.

lukasradek committed:
Extends G29 gcode parameters

Support G29 N and M parameters to set number of points in X and Y (overriding spacing) and/or G parameter to set spacing overriding number of points up to Config limit.

lukasradek commented:
I have added G29 parameters `N` (will have to rename), `M` and `G` to let user override the `GRID_MAX_POINTS_X` (and `Y`) and  `GRID_MIN_SPACING` configuration settings respectively.

G29 probing points determination logic:
Use `Configuration.h` values `GRID_MAX_POINTS_X`, `GRID_MAX_POINTS_Y` and `GRID_MIN_SPACING`

By default, it will keep the spacing between points at GRID_MIN_SPACING, but once the probing area gets big enough, that it would exceed the `GRID_MAX_POINTS_?`, it will cap it there and the spacing will get bigger.

If user specifies `N` or `M`, it acts as a "hard" override and it will probe that number of points in given dimension (up to the limit of `GRID_MAX_POINTS_?`).

If user specifies `G`, it acts as a "soft" override. It overrides `GRID_MIN_SPACING` but respects N or M parameters if set instead of `GRID_MAX_POINTS`.

thinkyhead committed:
make optional, start on leveling changes

thinkyhead commented:
The parameters `G` and `M` are verboten. The `N` parameter is also best avoided, though we sometimes begrudgingly allow it. If XY count parameters are needed, the usual choice is `I` and `J`.

lukasradek commented:
I have noticed that N is used ""internally" and now I am aware of the other two.
I will rename those.

thinkyhead commented:
Anyway, we already use `X` and `Y` parameters with LINEAR leveling to set the number of points in X and Y so that will simply be reused here with BILINEAR.

I went ahead and added a commit to…
- Make this behavior optional with a `VARIABLE_GRID_POINTS` option.
- Adjust the code that applies `GRID_MIN_SPACING`
- Remove the comments about `abl.abl_points` — Yes, that will also be recalculated, as it currently is with LINEAR, once the X/Y parameters are also applied to BILINEAR.
- Get a start on the changes in `bbl.h` and `bbl.cpp` which will need to apply to `get_z_correction`, `extrapolate_unprobed_bed_level`, etc., and the `ABL_BILINEAR_SUBDIVISION` methods.

thinkyhead committed:
'static' needed

thinkyhead committed:
extraneous parens

thinkyhead commented:
Other areas where this change will be taking effect include:
- EEPROM code in `settings.cpp` will need to save/load the number of points in X and Y.
- The `ABL_BILINEAR_SUBDIVISION` calculation of Catmull-Rom points will need to be updated.
- All LCD implementations that edit the mesh will need to be modified to account for variable size.

lukasradek commented:
Just to address a few thing that I noticed at first glance.
1) You removed the whole G29 parameter logic... what should happen with that?
2) You rewrote the variable grid points logic to be based on _spacing_ calculations instead of mine that was based on _nr of probe points_ calculation. I actually wrote it your way the first time, but then rewrote because it was a bit less calculations... is the way you wrote it better in some way?

thinkyhead committed:
GRID_MAX_CELLS…

thinkyhead commented:
> You removed the whole G29 parameter logic... what should happen with that?

You should reuse the logic already implemented for LINEAR but extend it to BILINEAR.

> You rewrote the variable grid points logic to be based on spacing calculations instead of mine that was based on nr of probe points calculation. I actually wrote it your way the first time, but then rewrote because it was a bit less calculations... is the way you wrote it better in some way?

I simply moved some lines around but tried to leave the logic the same. When calculations are made with `constexpr` values the compiler produces constants so there will be no extra calculation without `GRID_MIN_SPACING`. The only difference with and without `GRID_MIN_SPACING` is that there is additional calculation for `GRID_MIN_SPACING`, which applies with `VARIABLE_GRID_POINTS` or `AUTO_BED_LEVELING_LINEAR`.

Anyway, I'm done messing around and I'll get out of your way for a while as you work out the details. I assume you should be able to reuse the XY parameters that currently apply to LINEAR, but maybe you'll need to do it in a separate block. We'll see!

thinkyhead committed:
oops, ANY

thinkyhead commented on Marlin/src/gcode/bedlevel/abl/G29.cpp:
```diff
@@ -693,8 +727,8 @@ G29_TYPE GcodeSuite::G29() {
-          if (abl.verbose_level) SERIAL_ECHOLNPGM("Probing mesh point ", pt_index, "/", abl.abl_points, ".");
-          TERN_(HAS_STATUS_MESSAGE, ui.status_printf(0, F(S_FMT " %i/%i"), GET_TEXT(MSG_PROBING_POINT), int(pt_index), int(abl.abl_points)));
+          if (abl.verbose_level) SERIAL_ECHOLNPGM("Probing mesh point ", pt_index, "/", abl.grid_points.x * abl.grid_points.y, ".");
+          TERN_(HAS_STATUS_MESSAGE, ui.status_printf(0, F(S_FMT " %i/%i"), GET_TEXT(MSG_PROBING_POINT), int(pt_index), int(abl.grid_points.x * abl.grid_points.y)));
```
Here you should be able to keep using `abl.abl_points` — just make sure to calculate it as it is already calculated within the block for LINEAR that handles the X / Y parameters.

lukasradek replied:
I will now. I wasn't sure how that would play with `PROBE_MANUALLY`.

thinkyhead commented on Marlin/src/gcode/bedlevel/abl/G29.cpp:
```diff
@@ -109,6 +109,10 @@ class G29_State {
+     * If PROBE_MANUALLY is mutually exclusive with BILINEAR, abl_points is not needed when BILINEAR is active
+     */
     static constexpr grid_count_t abl_points = GRID_MAX_POINTS;
   #endif
```
Rather than adding more annoying commits I'll just add annoying comments. Here you can use…
```cpp
#if ANY(AUTO_BED_LEVELING_LINEAR, VARIABLE_GRID_POINTS)
  grid_count_t abl_points;
#elif ENABLED(AUTO_BED_LEVELING_3POINT)
  static constexpr grid_count_t abl_points = 3;
#elif ABL_USES_GRID
  static constexpr grid_count_t abl_points = GRID_MAX_POINTS;
#endif
```

thinkyhead committed:
fix my typos

thinkyhead commented:
Ok, just fixed my typos, and now I'll be scarce for a while! Hit me up on Discord if you run into any snags in any of the code affected by this option. You can go ahead and add `VARIABLE_GRID_POINTS` to one of the tests' (e.g., chitu_f103) `opt_set` section so that CI testing will flag any code issues as you go along.

thinkyhead committed:
just one more

lukasradek commented:
> * All LCD implementations that edit the mesh will need to be modified to account for variable size.

Are you sure about that? I wasn't aware that bilinear leveling would be displayed on screen somewhere (except the message "Probing x/total"). Now I have checked the `marlinui_DOGM.cpp` and I can see references to grid points only when drawing UBL interface, which is not our subject.

GMagician commented:
> I have noticed that `N` is used "internally" and now I am aware of the other two. I will rename those.

Industrial gcode uses N as a label to a line. Goto / loops may be done. I may say that it should be verboten as well as G and M

lukasradek committed:
Unifies LINEAR and BILINEAR logic, Implements EEPROM

Linear and Bilinear leveling now both support GRID_MIN_SPACING and variable grid points. G29 code should be done (with some testing ahead).

I have implemented EEPROM storage for grid dimensions, but it needs to be checked.

Added a Conditional so that GRID_MIN_SPACING is always defined (defaults to 0).

lukasradek committed:
Determine GRID_MAX_POINTS based on spacing

lukasradek commented:
**I have implemented the `G29.cpp` code to its probably functionally final form.**
Notable changes:
1. I am using `X`, `Y` and `P` for number of points per axis and `U` for spacing.
2. Previously if GCODE would supply incorrect number of points per axis, the leveling would just fail, which feels a bit harsh to me, since number of points does not affect safety if it is still constrained to the same area. So in my implementation it just silently caps the number of points to lower or upper limit.

I have fixed the mesh printing method in `bbl.cpp` so it can be used in `dryrun`.

I have attempted to implement the EEPROM writing/reading logic but that has to be checked by someone with deeper knowledge of that!

I have added to `Conditionals_post.h` definition of default value for GRID_MIN_SPACING in case that user comments it out. It defaults to 0, which means "any spacing is acceptable".

I have added a user comfort to `Conditionals_LCD.h`, that lets the user specify `GRID_MIN_SPACING` and comment out `GRID_MAX_POINT_[XY]`, which will be computed at compile time from bed size and spacing.

@thinkyhead Can you please check the EEPROM stuff and I have a question...
Is EEPROM storage supposed to survive a configuration change? It looks to me that the values retrieved simply by the order that they are in, meaning that if I configure Marlin differently, the order can change and therefore invalidate the EEPROM content. I am asking this because I was under assumption (possible witnessed) that Printing Stats in my printer survived config changes.

EvilGremlin commented:
> Is there any reason why SanityCheck forbids GRID_MAX_POINTS_[XY] to be less than 3?
Yes, textbook one - bilinear interpolatin/extrapolation can't work with 2.

lukasradek reviewed (commented):
@EvilGremlin Are you talking about interpolation or Catmull-Rom?
Because all I can see in the interpolation code (which I am talking about) is taking 4 corners of a square (2x2) and calculating some midpoint value.

lukasradek committed:
Fix constexpr and conditional compilations

EvilGremlin commented:
Sorry, i brainfarted and thought bicubic. Yeah, bilinear should work, but with 2x2 it's bound to produce wrong result (and damage bed). 
Linear is not mesh leveling, it just tilts the plane.

lukasradek commented:
@EvilGremlin Even if you probe just let's say 50 mm x 50 mm area?

EvilGremlin commented:
Depends on bed, but I'd say 25-30mm is go-to distance if you wait it to stick and look good. Anyway, 2x2 mesh just dont make sense: in every case, where it might be usable, 3point tilt will work just as fine if not better.

lukasradek committed:
Fix issue with printing array

lukasradek committed:
Set unused bedlevel z values to NAN

lukasradek committed:
ABL_SUBDIVISION - optimize and fix

Replace ad-hoc calculation of subdivided mesh size with precalculated value (just as it it for non-subdivided).

Fix subdivision nr of points calculation logic.

Fix printing of subdivision array.

lukasradek commented:
So i went through the `abl/bbl.cpp` and `abl/bbl.h` and this is the state of the affairs:

I replaced the logic where there were hardcoded MAX_POINTS (or equivalent) values with their runtime variants if `VARIABLE_GRID_POINTS` is enabled. So unless there is a code (especially in those mathematical algos) that relied on some other presumption (other than the previously hardcoded `GRID_POINT` limits), it should be all well and done.

I have done some testing and I am reasonably sure, that it is working with those exceptions:
1. `extrapolate_unprobed_bed_level` and  `extrapolate_one_point` should be checked, since I don't fully understand the algo and I don't have a machine with those properties.
2. Same goes for `line_to_destination`.
3. And even though I empirically tested `ABL_BILINEAR_SUBDIVISION` , someone should check `virt_2cmr`, `virt_cmr`, `virt_coord`, which I do not fully understand either.

As for possible suspects, before it was presumed, that the smallest grid will be 3x3, now it can be 2x2 - if the above mentioned algos relied on that... please fix 🙂.

lukasradek committed:
Enable in tests

lukasradek committed:
Fix G29 W, add ToDo

lukasradek committed:
G26, G42 supports VARIABLE_GRID_POINTS

Create new macros (GRID_USED_POINTS_[XY] etc.) that are substituted with correct runtime token to get the number of grid points currently used.

Preserve old macros (GRID_MAX_POINTS_[XY] etc.) so that they can still be used for the data array maintenance and by code that doesn't offer VARIABLE_GRID_POINTS (eg. UBL).

lukasradek committed:
M421 supports VARIABLE_GRID_POINTS

lukasradek committed:
M420 supports VARIABLE_GRID_POINTS

lukasradek committed:
Level Menu supports VARIABLE_GRID_POINTS + small enhancement

lukasradek committed:
Revert accidental platformio.ini change

lukasradek committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into G29-Variable-Grid-Point-Count

lukasradek committed:
Refactor

lukasradek committed:
AUTO_BED_LEVELING_LINEAR uses VARIABLE_GRID_POINTS macro

Linear leveling supports variable grid points by default, but the macro flags specified VARIABLE_GRID_POINTS only for bilinear leveling. That lead to a need to check ANY(AUTO_BED_LEVELING_LINEAR, VARIABLE_GRID_POINTS) everywhere for common variable grid points code.

AUTO_BED_LEVELING_LINEAR now enabled VARIABLE_GRID_POINTS by default (in Conditionals_LCD.h), so the abovementioned checks can be simplified to ENABLED(VARIABLE_GRID_POINTS), hence more semantic.

lukasradek committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into G29-Variable-Grid-Point-Count

thinkyhead committed:
Merge 'bugfix-2.1.x' into pr/26084

thinkyhead committed:
cleanup

thinkyhead committed:
No printf("%f") allowed

thinkyhead committed:
DISABLED

thinkyhead committed:
misc.

thinkyhead committed:
unused strings

thinkyhead committed:
clarify index

thinkyhead committed:
use xy_uint8_t

thinkyhead commented:
Progress is looking good. One of the more important displays to update is the Ender-3 V2 DWIN screen, especially with ProUI, which has one of the nicer mesh viewer/editor features.

thinkyhead committed:
ws

ThomasToka commented on Marlin/src/module/settings.cpp:
```diff
@@ -295,6 +295,9 @@ typedef struct SettingsDataStruct {
   xy_pos_t bilinear_grid_spacing, bilinear_start;       // G29 L F
   #if ENABLED(AUTO_BED_LEVELING_BILINEAR)
+    #if ENABLED(VARIABLE_GRID_POINTS)
+      xy_uint8_t bilinear_grid_points;
```
hi, where is this variable used?

lukasradek replied:
Hi,
it should be a place for `bedlevel.nr_grid_points`.

But I have to add that I wasn't familiar with how EEPROM storage works precisely so I worked by analogy.
I think I thought (😁) that the struct is there to define the structure of the EEPROM data and the variables themselves are not neccessarily supposed to be used.

ThomasToka commented on Marlin/src/module/settings.cpp:
```diff
@@ -1015,9 +1018,14 @@ void MarlinSettings::postprocess() {
 
       #if ENABLED(AUTO_BED_LEVELING_BILINEAR)
+        #if ENABLED(VARIABLE_GRID_POINTS)
+          EEPROM_WRITE(bedlevel.nr_grid_points);
```
Here you write bedlevel.nr_grid_points but later in 2041 "points" is loaded.

Trying to adopt your great work ;)

lukasradek replied:
Thanks 🙂.

`bedlevel.nr_grid_points` is a variable where the points are stored during runtime and other code can reach there and grab them.
`points` at line 2041 is a intermediary variable because the points cannot be loaded straight to the abovementioned variable but have to be processed by `bedlevel.set_grid(....)`

ThomasToka commented:
i have meanwhile done the same for my ubl builds and released it in my fork.

maybe you are interested:

https://github.com/ThomasToka/Marlin/commit/b4d1f86b60657b74380a7a7c6f610717722a2e87

ThomasToka committed:
Ubl variable grid points

Based on the PR https://github.com/MarlinFirmware/Marlin/pull/26084 i have done the same for UBL.
Todo: Integration in the display menues like in the original PR.
Needed configuration.h setting:
#define BLTOUCH
#define AUTO_BED_LEVELING_UBL
#define UBL_HILBERT_CURVE
#define Z_SAFE_HOMING
#define EEPROM_SETTINGS
#define GRID_MAX_POINTS_X 10  (this defines the biggest possible mesh)

ThomasToka committed:
Pass CI checks update

Encapsulate Conditionals_LCD.h and M420.cpp changes to pass CI build checks.

ThomasToka committed:
fix to one MESH_Y_DIST conversion

reviewing the code i have found falsely converted variable.

thinkyhead committed:
Pre-calculate grid distance

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26403

thinkyhead committed:
tweaks

thinkyhead committed:
etc

thinkyhead committed:
Init, load, store

thinkyhead committed:
fixes

thinkyhead committed:
etc

ThomasToka committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x-ubl-grid_points

thinkyhead committed:
set_grid_points

thinkyhead committed:
serial tweaks

thinkyhead committed:
more serial tweaks

thinkyhead committed:
move to work branch

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26084

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26403

thinkyhead committed:
Merge branch 'pr/26403' into pr/26084

thinkyhead committed:
merge followup

thinkyhead committed:
use var instead

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26084

vovodroid commented:
@EvilGremlin 
> with 2x2 it's bound to produce wrong result (and damage bed).

How much wrong? Just not precise as with more points (but good enough for small area), or it will be completely wrong calculated?

What about 2*(>=3) grid?

lukasradek commented:
> @EvilGremlin
> 
> > with 2x2 it's bound to produce wrong result (and damage bed).
> 
> How much wrong? Just not precise as with more points (but good enough for small area), or it will be completely wrong calculated?
> 
> What about 2*(>=3) grid?

That was a simplified and therefore misleading statement. It all depends on whether the state of your bed (probed area) can be described by just 2x2 points. Meaning there can only be linear tilt in each axis. 2 points in a single axis can describe linear gradient (slope) but cannot describe a wave (up, down, up). If you think about a large metal bed, it is unlikely that 2x2 points will be enough to produce a good mesh and can lead to a crash. But since this is a PR about variable grid points, the area that we are talking about can be as small as (for example) 5x5cm, which should be easily described by just 2x2 grid.

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26084

thinkyhead committed:
merge followup

vovodroid commented:
Hi, what is status of this PR?

lukasradek commented:
Just as you can see in the original post. I have done majority of the work and it is working for some specific HW configurations (from my view the most common, but I might be biased).

But it requires some math-capable person to check my implementation regarding the algos (should not be hard) and someone with e3v2 and ExtUI to implement it for those UIs. Even though I am not sure if that has to be mandatory for merging this PR.

Anyway as with some of my other PRs (Tramming Wizard, for example) I am not sure it the maintainers here have enough time to tend to these "nice-to-haves".

ThomasToka commented:
Thx @thisiskeithb. already did. Hope this is Ok for you @lukasradek . Will push the rest in the next days..

lukasradek commented:
It is fine by me, but beware that it has been quite a while for me and I don't really have enough time to get up to speed again with the system and changes that occured.

What is needed from me?

ThomasToka committed:
UBL TERN/OPTARG/GRID_LOOP_COND update

UBL TERN/OPTARG/GRID_LOOP_COND update
more conditionals to "VARIABLE_GRID_POINTS" and some TERN and Optarg

also GRID_LOOP_COND was introdudced to make the loop conditional, so it works without VARIABLE_GRID_POINTS like with it. it can be reused in bilinear.

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26084

ThomasToka commented:
Just thought about GRID_MIN_SPACING.

I think we should not include this here at this point and let it be calculated like it was before. Background of this is that it should be bound to the margins and let the user decide how near the points are.

```
spacing = 215 / (N - 1) ≥ 25
→ (N - 1) ≤ 215 / 25 = 8.6
→ N ≤ 9.6 → max integer N = 9
```
example with a 235x235 bed, 10 probing margin, 25 spacing, without taking probe offsets into account.
```
Mesh Size | X Spacing (mm) | Y Spacing (mm) | Valid?
3×3 | 107.50 | 107.50 | ✅
4×4 | 71.67 | 71.67 | ✅
5×5 | 53.75 | 53.75 | ✅
6×6 | 43.00 | 43.00 | ✅
7×7 | 35.83 | 35.83 | ✅
8×8 | 30.71 | 30.71 | ✅
9×9 | 26.88 | 26.88 | ✅
10×10 | 23.89 | 23.89 | ❌
11×11 | 21.50 | 21.50 | ❌
```
real world example with a Ender 3 S1 Pro 25 probing marginx, 42 probing margin y, 25 spacing, with Probeoffsets of -31.5x and -41.8y
```
Mesh Size | X Spacing (mm) | Y Spacing (mm) | Valid?
3×3 | 90.50 | 72.50 | ✅
4×4 | 60.33 | 48.33 | ✅
5×5 | 45.25 | 36.25 | ✅
6×6 | 36.20 | 29.00 | ✅
7×7 | 30.17 | 24.17 | ❌
8×8 | 25.86 | 20.71 | ❌
```
example with a 235x235 bed, 10 probing margin, no min spacing, without taking probe offsets into account.
```
Mesh Size | X Spacing (mm) | Y Spacing (mm) | Valid?
3×3 | 107.50 | 107.50 | ✅
4×4 | 71.67 | 71.67 | ✅
5×5 | 53.75 | 53.75 | ✅
6×6 | 43.00 | 43.00 | ✅
7×7 | 35.83 | 35.83 | ✅
8×8 | 30.71 | 30.71 | ✅
9×9 | 26.88 | 26.88 | ✅
10×10 | 23.89 | 23.89 | ✅
11×11 | 21.50 | 21.50 | ✅
12×12 | 19.55 | 19.55 | ✅
13×13 | 17.92 | 17.92 | ✅
14×14 | 16.54 | 16.54 | ✅
15×15 | 15.36 | 15.36 | ✅
```
another thing is that i would like to PR DYNAMIC_MARGINS as a follow up and with the i think GRID_MIN_SPACING would become more obsolete. with probe offsets the usable area is even smaller.

naturally one could set GRID_MIN_SPACING lower, but with the fact that we have probe offsets, a reachable area by the probe offsets and max and min positions. i think automatizing the spacing by those constants is a better way. otherwise the user would first need to understand his/her by the hardware given bounds and would not be able to use certain mesh sizes without poking with the settings by reducing GRID_MIN_SPACING and trying to set a bigger mesh size.

just my 2 cents to GRID_MIN_SPACING.

@lukasradek @thinkyhead what do you think?

lukasradek commented:
What is the issue with some of those variants ❌ not being valid?

ellensp commented:
@lukasradek those are build test run by CI 

"[Test mega2560] Azteeg X3 Pro | EXTRUDERS 5 | RRDFGSC | UBL | LIN_ADVANCE ......
In file included from Marlin/src/feature/bedlevel/bedlevel.h:67:0,
                 from Marlin/src/MarlinCore.cpp:183:
Marlin/src/feature/bedlevel/ubl/ubl.h: In static member function 'static float unified_bed_leveling::z_correction_for_x_on_horizontal_mesh_line(const_float_t, uint8_t, int)':
Marlin/src/feature/bedlevel/ubl/ubl.h:226:33: error: 'get_mesh_x' was not declared in this scope
     const float xratio = (rx0 - get_mesh_x(x1_i)) * TERN(VARIABLE_GRID_POINTS, mesh_dist_recip.x, RECIPROCAL(MESH_X_DIST));"

The show the PR is not compatible with common options

ThomasToka commented:
> The show the PR is not (yet) compatible with common options

I will follow up this in some minutes

<img width="1920" height="1080" alt="grafik" src="https://github.com/user-attachments/assets/c5b41ea9-0baa-4878-a7a3-7007286acc1c" />

ThomasToka committed:
UBL TERN/OPTARG/GRID_LOOP_COND update followup

ubl.cpp ubl.h
fix checks

ThomasToka commented:
> Can you elaborate? 🙂

does this not reduce the grid_points when it does not fit with the given GRID_MIN_SPACING.
but maybe i am wrong.

lukasradek commented:
Yes, as it should. That is the whole point of variable grid points, that you probe only as many as it is necessary. There is no point in probing 10x10 on a 10x10 mm print area (if we go to extremes).

ThomasToka commented:
Another thing is that lcd/tft sometimes have firmware with hardcoded sites, so the fallback site must also exist. This means every display that has extra sites for the mesh sizes needs to have not only the supported sites but also the fallback sites. and as this is trigerred/calculated by the GRID_MIN_SPACING the fallback site could be 7, 6 or 5 for example. Adds a lot of logic on the display site for a very "not expected outcome" of the mesh set process.

lukasradek commented:
If you don't want addition layer of calculation or don't want the printer to adapt the mesh size, why would you turn the variables grid points option on? Furthermore user is not choosing the mesh size. He chooses the maximum mesh size. And he couldn't care less if the printer actually chooses to probe smaller number of points as long as it creates a reliable mesh.

ThomasToka commented:
Please understand that lcd/tft send the result of the probe to a hardcoded adress on the display. Where a base address is incremented from the start address. Also after mesh change a new display site is triggerred. And this needs to exist.

There is really no chance to adopt this approach for a wide range of the screen firmwares of DGUS/DWIN/TJC whatever displays. All would need a update. Most Displays i know have very low space left. Without this one simply pics those sites in the lcd/tft class that are supported.

lukasradek commented:
I don't understand why would mesh leveling need to send anything to the displays' firmware. You would have to explain.

Also what feature-set do you expect Variable grid points to offer when you don't allow for runtime calculation of mesh size?

ThomasToka commented:
<img width="1920" height="1080" alt="grafik" src="https://github.com/user-attachments/assets/5dced452-ae8e-4cea-ad73-b221656574fd" />
<img width="1920" height="1080" alt="grafik" src="https://github.com/user-attachments/assets/ed4aeda4-ff23-4054-8d57-9a1dd7671b91" />

and then you do on mesh change the following for the display:

as a example from my lcd class when i change grid_used_points:

```

void RTS_SetGridMaxPoints(uint8_t gridmaxpoints, uint8_t m19load)
{
  if (gridmaxpoints == 5 || gridmaxpoints == 7 || gridmaxpoints == 10){
    RTS_ResetMesh();
    lcd_rts_settings.max_points = gridmaxpoints;
    RTS_LoadMeshPointOffsets();
    RTS_SetBltouchHSMode();
    rtscheck.RTS_SndData(lcd_rts_settings.max_points * lcd_rts_settings.max_points, AUTO_BED_LEVEL_END_POINT);
    rtscheck.RTS_SndData(gridmaxpoints, SET_GRID_MAX_POINTS_VP);
    bedlevel.max_points.x = gridmaxpoints;
    bedlevel.max_points.y = gridmaxpoints;
    if (settingsload != 1) queue.enqueue_now_P(PSTR("M84"));
    queue.enqueue_now_P(PSTR("G92.9Z0"));
    if (m19load == 0){
      rtscheck.RTS_ChangeLevelingPage();
    }
    RTS_ShowMotorFreeIcon(true);
  }
} 
```
so you compile for the max possible by your display, and then pic those that your display supports.

hope you understand now more..

ThomasToka committed:
ABL/UBL TERN/OPTARG/GRID_LOOP_COND update

Update more of the files to be conditional to VARIABLE_GRID_POINTS.

Touched some display files but this is ok, when GRID_LOOP_COND is available.

Update Conditionals-3-etc.h to fix UBL compile

lukasradek commented:
Thanks.

Without deep knowledge of DWIN and the lcd/tft issues, let's stick to the following:

1. Variable number of grid points will not be dropped from this PR even if some HW might not support it. It is the whole point of this feature.

2. For devices where the support is problematic, we should follow this protocol:
2.1. Can it be implemented even if not easily solely in Marlin?
2.2. If not? Is it neccessary to preserve all of the previous UI? We are concerned with mesh leveling that the printer does in order to lay the first layer properly. I still don't see why a user should be bothered with this in UI.
2.3. If it is? Is there a way to do it suboptimally? If the UI supports 10x10 mesh and has enough hardwired stuff to support this, there shouldn't be a reason why any smaller mesh shouldn't work. And the UI will just leave the unused points unpopulated.
2.4. If for some reason even that cannot be done? Too bad, Variable Grid points is not supported for those devices.

lukasradek commented:
Ok, let's say we ditch GRID_MIN_SPACING. How many points do you expect the printer to probe when GRID_MAX_POINTS_X and GRID_MAX_POINTS_Y will be set to 10 and print job comes in that uses only 50 x 50 mm on the print bed?

ThomasToka commented:
Attached the script i used for the visualisation if you want to check other combinations.

[meshes.txt](https://github.com/user-attachments/files/21326976/meshes.txt)

ThomasToka commented:
I think you only look at your own workflow by using gcodes from a terminal.
I look more to have the whole system dynamic and usable -also- with LCD controllers those are very limited on their possibilities.
Most random users dont even know what a terminal is and how this all works.

So please dont be mad at me. I want to understand everything and make it usable for a wide range of printers/lcd/tft .

ThomasToka commented:
@vlsi what is your problem? you have a lcd?

ThomasToka commented:
would this not fix it?
https://github.com/lukasradek/Marlin/blob/dbd46a423f8553c906a21fcebafd5f532b6aae65/Marlin/src/gcode/bedlevel/abl/G29.cpp#L483
```
if (parser.seenval('Y') && parser.seen('X')) {
...
}

or even
if (parser.seenval('Y') || parser.seen('X')) {
...
}
```
so a plain "G29" is not affected?

vlsi commented:
@ThomasToka , could you please clarify what do you mean by "dynamic margins" and could you please clarify who and when should supply those "dynamic margins"?

Frankly speaking, I agree with @lukasradek: the parameters should have physical meaning.
When it comes to bed probing, `GRID_MIN_SPACING` has a clear meaning. For instance, if bed hills take 40mm to raise and then 40mm to fall (e.g. in X direction), then user might assume it would be enough to sample the bed height with 20mm intervals. See https://en.wikipedia.org/wiki/Nyquist_frequency.
The number of samples (e.g. 5x5 grid or 10x10 grid) is not relevant to capture the bed surface. The frequency (==the distance between the samples) of the samples is important. `GRID_MIN_SPACING` provides exactly that: it allows users to configure the minimum distance between the samples. 

---

> Exactly 10. when you use only 50x50mm the right way would be to set a Probing margin of 80 for example. so you have 235-80-80= 75x75 in the middle

@thinkyhead , can you please clarify what was the reason to have 10x10 grid (100 samples) over 50x50mm area? Frankly, I would consider it to be a waste of time in most of the cases. As a end-user, I do not care about the number of points in the grid. I do care about the distance between the sample points.

In other words, as a end-user, I would probably prefer configuring a "distance between bed probe samples" rather than the "number of points in a grid". I would prefer to sample only the part of the bed that is relevant for the first layer (which varies from print to print), and I would prefer the machine to select the number of samples based on the desired distance between the probe samples. 

---

>you have a lcd?

I've TFT with ColorUI. I'm not sure why does that matter though. Sure it would be nice to display the bed map, however, I think the UI support is optional. If the UI can't handle the dynamic number of points for some reason, then the corresponding screen could be hidden from the UI, and the user could configure bed leveling with menu-style screens.

lukasradek commented:
> 
> > Exactly 10. when you use only 50x50mm the right way would be to set a Probing margin of 80 for example. so you have 235-80-80= 75x75 in the middle
> 
> @thinkyhead , can you please clarify what was the reason to have 10x10 grid (100 samples) over 50x50mm area? Frankly, I would consider it to be a waste of time in most of the cases. As a end-user, I do not care about the number of points in the grid. I do care about the distance between the sample points.
>

That was my extreme example just to prove that min spacing is necessary to prevent useless probing.

I think @ThomasToka has has an issue with something else that he is trying to incorrectly solve with messing with this part of the system. Reminds me of a guy who was trying to solve his undertensioned extruder idler by increasing the flow rate 😃. It sort of might work but it breaks other things.

ThomasToka commented:
> I think @ThomasToka has has an issue with something else that he is trying to incorrectly solve with messing with this part of the system. Reminds me of a guy who was trying to solve his undertensioned extruder idler by increasing the flow rate 😃. It sort of might work but it breaks other things.

I am not the one that misses here a disrupting part on the other side of the code here.

And then you start to be disgusting to me. I have a fully dynamic system with thousands of production users. Do you have?

So please stay professional.

@vlsi please have a look in this commit on top of the actual bugfix-2.1.x:  https://github.com/ThomasToka/Marlin/commit/03eb86c275fedc8caa2f449bd8c2e4796eb591ca#diff-52ecf1940e83aa7430b35b9edd0fec5fe2b1e61d565823492de1382ced75b817 
so i will deliver dynamic margins, dynamic probecount when we have this here merged. everything conditional. so users not using the dynamic compile can use marlin like they did before.
if you understand then you understand. i not then ask what you not understand.

vlsi commented:
@ThomasToka , I still do not understand who and when should supply those "dynamic margins".
Frankly, the commit is `+12589 -862` and it is not clear what exact part do you mean.

For instance, one of the use cases is to "probe the surface under the first layer of the print".
Then, if a print is just 50x50mm, then only a few probe samples would be enough (e.g. 3x3 grid with 25mm spacing).
If a print is 100x100mm, then it would be like 5x5 grid with the same 25 mm spacing.

Can you please clarify how it would be covered by the suggested "dynamic margins"? Do you mean the user should supply the number of grid points for every G29 command? If so, it would be inconvenient as one of the parameters (==number of grid points) would depend on another parameter (==probe area size).

ThomasToka commented:
> @ThomasToka , I still do not understand who and when should supply those "dynamic margins". Frankly, the commit is `+12589 -862` and it is not clear what exact part do you mean.
> 
Kindly have a look here:
probe gets the desired margin:
https://github.com/ThomasToka/Marlin/blob/03eb86c275fedc8caa2f449bd8c2e4796eb591ca/Marlin/src/module/probe.h#L110
define of defaults (later moved out of the lcd class to a global variable that can be reused by all implementations:
https://github.com/ThomasToka/Marlin/blob/03eb86c275fedc8caa2f449bd8c2e4796eb591ca/Marlin/src/lcd/rts/e3s1pro/lcd_rts.cpp#L281
setting margins (later also moved to a global function):
https://github.com/ThomasToka/Marlin/blob/03eb86c275fedc8caa2f449bd8c2e4796eb591ca/Marlin/src/lcd/rts/e3s1pro/lcd_rts.cpp#L5038
> 
> For instance, one of the use cases is to "probe the surface under the first layer of the print". Then, if a print is just 50x50mm, then only a few probe samples would be enough (e.g. 3x3 grid with 25mm spacing). If a print is 100x100mm, then it would be like 5x5 grid with the same 25 mm spacing.
> 
With dynamic margins you reduce insrease the margin, and as you also have dynamic/variable grid points you can lower them also. But this should be the decision of the user.
> 
> Can you please clarify how it would be covered by the suggested "dynamic margins"? Do you mean the user should supply the number of grid points for every G29 command? If so, it would be inconvenient as one of the parameters (==number of grid points) would depend on another parameter (==probe area size).

Nope. 

Looks on my UI like this then:

<img width="452" height="603" alt="292521030-11f99d32-a33f-497b-9f9c-d68c32b7d53c" src="https://github.com/user-attachments/assets/f0c8e25e-edcd-4f13-87de-9792e5de67f3" />
All blue fields are editable without reboot on runtime, probes 1-5, mesh 5,7,10, margin x, margin y
Also it reaccounts the min possible margin under accounting of the probe offsets if you enter "0" as a margin. both x and y independeant.

In my implementation still a G29 and also a G29 L50 R150 F50 B150 V4 from terminal is possible and the user does exactly what he wants and probe the area he/she/it wants. naturally a user does not probe a 50x50 print peace with 10x10 points and a low margin. if a reduction of probing area is wanted you choose margin x 80 margin y 80, 5x5, and if your bed is stiff and it should be fast you only choose one probe at a point. all dynamic/variable.

vlsi commented:
> you also have dynamic/variable grid points you can lower them also. But this should be the decision of the user.

I think this is what `GRID_MIN_SPACING` is about.
The user should **not** care about the grid size. If you think there's a use case when the grid size is important, please spell it aloud. I truly do not understand why would one want configuring the number of points in a grid unless they have very tight memory constraints. The grid size could be fully automatic, and the user should not care to adjust the number of grid points from prints to prints. At the same time, the firmware should be smart enough to use 1x1 or 3x3 grid for 50x50mm print job to reduce the print duration.

> With dynamic margins you reduce or increase the margin

Frankly, I do not care the way I configure probing area. If that goes with "dynamic margins", that's fine. If probing area comes as G29 parameters, that's fine as well. I think "dynamic margins" is a completely different discussion.

"dynamic margins" have absolutely nothing to do with "dynamic number of grid points", so could you please leave "dynamic margins" alone when discussing `GRID_MIN_SPACING`?

`GRID_MIN_SPACING` is there so the firmware automatically selects the appropriate number of grid points, so the user does not have to guess between 3x3 5x5 and so on.

Based on your [comments](https://github.com/MarlinFirmware/Marlin/pull/26084#issuecomment-3088943184), it looks like you think `GRID_MIN_SPACING` should be removed. However, it looks like you suggest **no** alternative option to automatically select the number of grid points. It looks like you suggest that the user should manually select the number of grid points every time, however, it is not clear why require a manual step when it could be automatic, and it is not clear how the user would pick the grid size either.

vlsi commented:
> Do never change anything on runtime that a user not wants. And thats exactly what happens here

Frankly, it looks like the culprit. You think that "if the user selects 5x5, then machine should follow the exact command and build 5x5 grid". I would strongly disagree with that when it comes to "grid size".

Unfortunatley, you do not provide the reason the grid size is important. Please, hear me out: the grid size is **not** important.
I'm all ears to hear about a case when the grid size it important, yet you provide none. What is the practical meaning in the "number of points in a grid"? I truly do not understand why do you think the machine should follow the exact number of points in a grid.

The user cares about having "good enough mesh" and the user cares to spend "the minimum amount of time for building a good mesh". They should not care about the exact number of grid points provided the resulting mesh is good enough.

> ThomasToka: but are punished if you have a big mesh set to 10x10 and use a plain G29 with the reduction of the mesh size.

I think you misunderstand `GRID_MIN_SPACING`. It reduces the number of grid points only if the probe points become too close which means reducing the grid size keeps the grid quality and it reduces the time it takes to build the mesh. It does not reduce the points at random. It just follows https://en.wikipedia.org/wiki/Nyquist_frequency.

ThomasToka committed:
Ternary VARIABLE_GRID_POINTS for displays

Populated TERN(VARIABLE_GRID_POINTS, GRID_USED_POINTS[_][X|Y], GRID_MAX_POINTS[_][X|Y]) where needed to allow the usage of this PR.
constexpr disabled where needed when VARIABLE GRID_POINTS is used.

ThomasToka committed:
UBL unify with ABL

grid_points -> nr_grid_points

lukasradek commented:
It was working before, the issue was that it received no attention for about a year.

ThomasToka commented:
it’s still working. as soon as the rest is committed. maybe there will be some tweaks needed for some displays, but i looked in every display and touched  only changed where it’s logically needed.

ThomasToka commented:
@ellensp the fixes needed are in the PR i mentioned. i have seen what causes it but cant commit it directly to this PR.

ThomasToka commented:
we need some short @thinkyhead action here for the merge.. (sorry for tagging you scott)

ThomasToka commented:
@ellensp will it be automatically added to this PR when @lukasradek aproves my PR in his branch?

lukasradek commented:
> @lukasradek The PR needs to pass basic CI tests before they will even be considered for merging.
> 
> This code does not pass the automated CI tests. Currently this code will break compiling Marlin for a large portion of users.

Oh... so it is almost 2 years now.

I implemented this PR around September 2023 and it was passing checks no problem. For I while I even kept it in sync with the master, so it can be merged easily, but noone really did anything with it, so I stopped.

The current broken state is the result of Thomas's commit, but any previous commit could be used.

If the maintainers want this in production Marlin, I would be willing to put some more work into it, but I need some instruction on what is neccessary, so it doesn't get stale like few years ago.

ThomasToka commented:
the final remerge/tweaking is done by thinkyhead. so for now is enough to keep it passing checks with its state. and that will be the case when the PR 3 is in here. then most of your checkmarks on your todo are done.

lukasradek commented:
> @lukasradek go to your [lukasradek#3](https://github.com/lukasradek/Marlin/pull/3)
> 
> scroll down and click squash and merge to add those patches to your code

I most likely need that branch to stay as it is for now since I am using it for my printer and don't really want to commit changes that I haven't gone through. 

Any reason why he cannot modify this PR directly? It seems that he already successfully made one commit to it before.

ellensp commented:
You cannot have it both ways.   
You created the pr from the G29-Variable-Grid-Point-Count  branch so that branch that needs updated.. to conform to required marlin coding. 

Create a new branch "G29-Variable-Grid-Point-Count-myprinter" from G29-Variable-Grid-Point-Count   and use that for your printer, then merge  the changed to G29-Variable-Grid-Point-Count

ThomasToka commented:
you can/should create a new branch in your github desktop from the desired commit of your branch. probably better. i will surely have to follow up so we are fully TERNary with it and we can continue to make this mergable.

lukasradek commented:
@ellensp I regularly see other people (for example thinkyhead) commiting into this PR without my "approval". What is different in this case?

ellensp commented:
"On user-owned forks, if you want to allow anyone with push access to the upstream repository to make changes to your pull request, select Allow edits from maintainers."

This is the default 

Ie this gives us MarlinFirmware maintainers permissions to edit your fork the PR came from.

But these permissions seem to have been removed.

lukasradek commented:
<img width="421" height="1107" alt="image" src="https://github.com/user-attachments/assets/34f398b3-fdc3-4798-891e-ef2ee3e98f85" />

Thanks for the explanation. I was lightly aware of this setting but wasn't sure on what level does it apply. Anyway it was checked the whole time, so I am not sure where is the issue. I have re-checked it in case it might affect something.

lukasradek committed:
Merge pull request #3 from ThomasToka/G29-Variable-Grid-Point-Count

ABL/UBL TERN/OPTARG/GRID_LOOP_COND update followup

Merging on good faith without any checks what the changes are.

ThomasToka committed:
ABL/UBL/Settings pass checks

bbl.h
ubl.h
ubl_G29.cpp
settings.cpp

ellensp commented:
on your local machine run make

make tests-single-local        : Run a single test locally
make tests-single-local-docker : Run a single test locally, using docker
make tests-all-local           : Run all tests locally
make tests-all-local-docker    : Run all tests locally, using docker

lukasradek commented:
@ThomasToka 
I have seen even you commit directly to this PR and for some reason you are marked as Reviewer here. There should be no reason why I would have to be the middleman now.

<img width="1545" height="376" alt="image" src="https://github.com/user-attachments/assets/fc333805-2edc-422d-b236-f0a4dfc90a69" />

ThomasToka commented:
Those wer brought over by thinkyhead. I did a PR against Marlin. Thinky merged them to your PR as it is the same for UBL.
Sorry again. I afaik can not change this. :-(

ThomasToka commented:
i am right now setting up my local installation to be able to do all tests local myself.

this will result in a pull request for the docker image as it is not usable in its actual (buster) state. i am right now so far that i can run tests. those errors were simply not visible not doing all checks. sorry for this. i will fix it all till it matches.

i will come back tomorrow or later with another PR to Lukas´Branch so we finally can fix the checks.

and with the PR of the docker image ^^ working now on the failed checks already.

ThomasToka commented:
finally. i just PRed the docker fix if one wants to use the checks locally.

i also already have written the fix for this but will first run the checks over it. it takes forever ^^

but first checks are ok now. i will let it run over night, and have a look tomorrow morning.
```
[Test linux_native --target clean] Setup Build Environment...
Platform Manager: Installing native
Downloading 0% 10%
Unpacking 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%
Platform Manager: native@1.2.1 has been installed!
Passed

[Test linux_native] Linux with EEPROM...
Marlin/src/inc/Warnings.cpp:118:4: warning: #warning "Your Configuration provides no method to acquire user feedback! (Define NO_USER_FEEDBACK_WARNING to suppress this warning.)" [-Wcpp]
  118 |   #warning "Your Configuration provides no method to acquire user feedback! (Define NO_USER_FEEDBACK_WARNING to suppress this warning.)"
      |    ^~~~~~~
Passed
All tests completed successfully
Running tests for simulator_linux_release
Running simulator_linux_release Tests

[Test simulator_linux_release --target clean] Setup Build Environment...
Passed
Fetching examples/Simulator configurations from bugfix-2.1.x...

[Test simulator_linux_release] Simulator...
lto-wrapper: warning: using serial compilation of 31 LTRANS jobs
lto-wrapper: note: see the ‘-flto’ option documentation for more information
Passed
All tests completed successfully
Running tests for mega2560
Running mega2560 Tests

[Test mega2560 --target clean] Setup Build Environment...
Platform Manager: Installing atmelavr @ ~4.0.1
Downloading 0% 10% 20% 30% 40% 50% 60%
Unpacking 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%
Platform Manager: atmelavr@4.0.1 has been installed!
Passed

[Test mega2560] Default Configuration | no source filtering...
Tool Manager: Installing platformio/toolchain-atmelavr @ ~1.70300.0
Downloading 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%
Unpacking 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%
Tool Manager: toolchain-atmelavr@1.70300.191015 has been installed!
Tool Manager: Installing platformio/framework-arduino-avr @ ~5.1.0
Downloading 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%
Unpacking 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%
Tool Manager: framework-arduino-avr@5.1.0 has been installed!
```
and then PR the needed changes again sometime tomorrow around 10CET.

lukasradek committed:
Merge pull request #4 from ThomasToka/G29-Variable-Grid-Point-Count

Merging on good faith. Haven't checked the code thoroughly.


bbl.h Fix

ThomasToka committed:
UBL update ui_api.cpp

Fix ui_api check

ThomasToka commented:
Thanks for merging. Sadly i still cant run all tests locally at once. So i am dependant on the online checks.
But i am also working on a fix for the run_tests scripts so they are more reliable
I put another fix for the ui_ap.cpp check. https://github.com/lukasradek/Marlin/pull/5 Sorry again.

What it does:
- add your missing todos
- make it fully TERNary so it can be merged without interruption for other implementations
- simplyfy some code
- make UBL the same like BILINEAR (i will have to add your spacing as last, thats the only difference atm)

Thank you so far for the merges!

lukasradek committed:
Merge pull request #5 from ThomasToka/G29-Variable-Grid-Point-Count

UBL update ui_api.cpp

Merging on good faith without code evaluation.

ThomasToka commented:
Thanks for merging. We are now fine with the checks. Good idea with the fork. But i think it is a bit more of work needed to use the tests on github. But i will have a look into this.

The tests are now fine. I will write a little more tomorrow to your questions as i yesterday felt down the stairs and my right foot is demaged. I will go now to the surgery to make "a pic" of my leg ^^

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26084

thinkyhead committed:
merge followup, more conditionals

thinkyhead commented on Marlin/src/lcd/extui/dgus_reloaded/config/DGUS_Addr.h:
```diff
@@ -26,7 +26,7 @@
 #define DGUS_FILENAME_LEN     32
 #define DGUS_ELAPSED_LEN      15
-#define DGUS_LEVEL_GRID_SIZE  25
+#define DGUS_LEVEL_GRID_SIZE  GRID_VAL(GRID_USED_POINTS, 25)
```
The grid code for some displays is hard-coded to the manufacturer's specifications. Since not all displays will accomodate all forms of ABL, either, we may need to add sanity checks about which displays can be used with this new feature.

thinkyhead committed:
clean up and fix this and that

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26084

thinkyhead committed:
use GRID_CONSTEXPR

thinkyhead committed:
Add DEBUG_ECHO_NUM

thinkyhead committed:
finis
